Cleanup commands and quadtree

peppsac opened the pull request:
Note 1: first commit is from PR #94 
Note 2: do not merge yet as this introduces a regression when zooming quickly (tile blinking)

peppsac committed:
Rename Camera.SSE -> Camera.computeNodeSSE

The new name reflects what the method really does.
Also, this doesn't change node's sse value anymore.

peppsac committed:
Move InterfaceCommander creation from Layer to Quadtree

Quadtree is the only Layer with a valid interCommand; others have interCommand==undefined.
So it makes sense to  move InterfaceCommander to Quadtree constructor.

peppsac committed:
Allow custom command priority functions

There is no 'one size fits all' priority function, so we let
the users of InterfaceCommander define their own.

For now the sole user of InterfaceCommander is Quadtree, so the
old priority function has been moved there.

peppsac committed:
Replace Quadtree's command priority function

Priority:
 * subdivision
 * uncomplete tiles
 * tiles meant to be displayed
 * others

The priority doesn't use sse or displayed attributes but instead prioritize
by category of command.
'sse' values for instance are meaningless as soon as the camera move/zoom so
comparing commands sse values doesn't help.

peppsac committed:
Prioritize recents commands

Command priority is computed at command creation so this commit adds
a way to prioritize recent commands over old ones (if they have the same
priority).

peppsac commented:
Rebased on master.

Jeremy-Gaillard commented on src/Scene/Quadtree.js:
```diff
@@ -12,14 +12,39 @@
+], function(Layer, InterfaceCommander, Quad, NodeMesh) {
+
+
+    function commandQueuePriorityFunction(cmd) {
```
I would move this function to NodeProcess since the priority of a command depends more on the nature of the node than the structure in which it is.
Edit: just noticed this would also require passing the priority value directly to the interfaceCommander instead of using a function

peppsac replied:
Makes sense but I would keep it in `Quadtree` until we have another implementation (Octree maybe?) that uses NodeProcess and this priority function.
WDYT?

gchoqueux commented:
41e2d96447c2013390a117acc3551c580f010801, the policy priorities needs thinking.
There's not trivial solution and there are lot of parameters : subdivision, request texture, visibility, point of interest, anticipation of movement. Ideally, a user can customize this priorities (like c5db57ad0cc18bc31d2235300fd3d3a85efef05e). I add issue  #136

peppsac commented:
> the policy priorities needs thinking.

Agreed. The changes proposed here are a first step in this direction.

gchoqueux commented:
r- remove regression when zooming quickly (tile blinking)

peppsac committed:
Make sure children tiles are all ready when hiding a parent

Currently only Quadtree based subdivision is supported so we simply need
to make sure that a parent node has 4 children and that they're all loaded.

peppsac commented:
@gchoqueux blinking issue is fixed (see last commit). r?

gchoqueux commented:
r+
